libpages/server.go

```diff
@@ -121,12 +167,56 @@ func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 			msg:    "CtxWithRandomIDReplayable",
 			logger: s.config.Logger,
 		})
-	fs, err := s.getFS(ctx, root)
+	st, err := s.getSite(ctx, root)
+	if err != nil {
+		s.handleError(w, err)
+		return
+	}
+
+	cfg, err := st.getConfig(false)
+	if err != nil {
+		// TODO: error page to show the error message?
+		s.config.Logger.Info("getConfig", zap.String("host", r.Host), zap.Error(err))
+		s.handleError(w, err)
+		return
+	}
+
+	var canRead, canList bool
+	var realm string
+	user, pass, ok := r.BasicAuth()
+	if ok && cfg.Authenticate(user, pass) {
+		canRead, canList, realm, err = cfg.GetPermissionsForUsername(
+			r.RequestURI, user)
+	} else {
+		canRead, canList, realm, err = cfg.GetPermissionsForAnonymous(
+			r.RequestURI)
+	}
 	if err != nil {
 		s.handleError(w, err)
 		return
 	}
-	http.FileServer(fs).ServeHTTP(w, r)
+
+	// Check if it's a directory containing no index.html before letting
+	// http.FileServer handle it.  This permission check should ideally
+	// happen inside the http package, but unfortunately there isn't a
+	// way today.
+	isListing, err := s.isDirWithNoIndexHTML(st, r.RequestURI)
```

Comment:
As above, won't `r.RequestURI` include the domain info?

Reply:
No I think it's the path under the domain. So for `https://keybase.io/jzila`, `RequestURI` would be `/jzila`

Reply:
For example, for a request to `https://kbp-test.song.gao.io/`, this is what we get in logs:
```
2017-12-23T02:13:48.226Z        INFO    libpages/server.go:147  ServeHTTP       {"host": "kbp-test.song.gao.io", "urk": "/", "proto": "HTTP/2.0"}
```

Reply:
The RFC says it can be the absolute URI:
https://www.w3.org/Protocols/rfc2616/rfc2616-sec5.html#sec5.1.2

Reply:
Seems like what you want is `r.URL.Path`

Reply:
Oh good catch. And it seems the RequestURI can include parameters as well. Will do; thanks!